pymtl3/dsl/Component.py

```diff
@@ -55,10 +61,542 @@ def _construct( s ):
+
+      list_parent = getattr( parent, name )
+      i = 0
+      while i < len(indices) - 1:
+        list_parent = list_parent[ indices[i] ]
+        i += 1
+      list_parent[ indices[i] ] = obj
+
+      # NOTE THAT we cannot use setattr hook for the top level metadata
+      # because setting an element in the list doesn't call setattr.
+      # As a result we have to copy some code from setattr hook and do it
+      # manually here for the top level
+
+      obj._dsl.parent_obj = parent
+      obj._dsl.level      = parent._dsl.level + 1
+      obj._dsl.my_name    = u_name = name + "".join( [ "[{}]".format(x)
+                                                     for x in indices ] )
+      # FIXME properly implement param_tree
+      # Iterate through the param_tree and update u
+      if parent._dsl.param_tree is not None:
+        if parent._dsl.param_tree.children is not None:
+          for comp_name, node in parent._dsl.param_tree.children.iteritems():
+            if comp_name == u_name:
+              # Lazily create the param tree
+              if obj._dsl.param_tree is None:
+                obj._dsl.param_tree = ParamTreeNode()
+              obj._dsl.param_tree.merge( node )
+
+            elif node.compiled_re is not None:
+              if node.compiled_re.match( u_name ):
+                # Lazily create the param tree
+                if obj._dsl.param_tree is None:
+                  obj._dsl.param_tree = ParamTreeNode()
+                obj._dsl.param_tree.merge( node )
+
+      obj._dsl.full_name = ( parent._dsl.full_name + "." + u_name )
+
+      # store the name/indices
+      obj._dsl._my_name     = name
+      obj._dsl._my_indices  = indices
+
+      NamedObject.__setattr__ = NamedObject.__setattr_for_elaborate__
+      obj._construct()
+      del NamedObject.__setattr__
+
+    top = s._dsl.elaborate_top
+
+    added_components = obj._collect_all( [ lambda x: isinstance( x, Component ) ] )[0]
+
+    # First elaborate all functions to spawn more named objects
+    for c in added_components:
+      c._dsl.elaborate_top = top
+      c._elaborate_read_write_func()
+
+    added_signals, added_method_ports = \
+      obj._collect_all( [ lambda x: isinstance( x, Signal ), \
+                          lambda x: isinstance( x, MethodPort ) ] )
+
+    top._dsl.all_components    |= added_components
+    top._dsl.all_signals       |= added_signals
+    top._dsl.all_method_ports  |= added_method_ports
+
+    top._dsl.all_named_objects |= added_components
+    top._dsl.all_named_objects |= added_signals
+    top._dsl.all_named_objects |= added_method_ports
+
+    for c in added_components:
+      top._collect_vars( c )
+
+    for c in added_signals | added_method_ports:
+      c._dsl.elaborate_top = top
+
+    # Lazy -- to avoid resolve_connection call which takes non-trivial
+    # time upon adding any connect, I just mark it here. Please make sure
+    # to call s.get_all_value_nets() to flush all pending connections
```

Comment:
I think the pending connections will be resolved once you need it right? The comment here sounds like your have to manually call some API to resolve connections before using them...

Reply:
Good point. The comment is copied from the old implementation. What I actually mean is that whenever you call get_all_value_nets which is the API that gives you all the nets, the API itself will check if there are pending connections and flush them to return the latest nets.